elkodon/src/sample_mut.rs

```diff
@@ -64,38 +65,65 @@ impl<
         Service: service::Details<'config>,
         Header: Debug,
         MessageType: Debug,
-    > SampleMut<'a, 'publisher, 'config, Service, Header, MessageType>
+    > SampleMut<'a, 'publisher, 'config, Service, Header, MaybeUninit<MessageType>>
 {
     pub(crate) fn new(
         publisher: &'publisher Publisher<'a, 'config, Service, MessageType>,
-        ptr: NonNull<MaybeUninit<Message<Header, MessageType>>>,
+        ptr: RawSampleMut<Header, MaybeUninit<MessageType>>,
         offset_to_chunk: PointerOffset,
     ) -> Self {
         publisher.loan_counter.fetch_add(1, Ordering::Relaxed);
+
+        // SAFETY: the transmute is not nice but safe since MaybeUninit has the same layout as the inner type
+        let publisher = unsafe { std::mem::transmute(publisher) };
+
         Self {
             publisher,
             ptr,
             offset_to_chunk,
         }
     }
 
+    /// Extracts the value of the `MaybeUninit<MessageType>` container and labels the sample as initialized
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that `MaybeUninit<MessageType>` really is initialized. Calling this when
+    /// the content is not fully initialized causes immediate undefined behavior.
+    pub unsafe fn assume_init(
+        self,
+    ) -> SampleMut<'a, 'publisher, 'config, Service, Header, MessageType> {
+        // the transmute is not nice but safe since MaybeUninit has the same layout as the inner type
```

Comment:
As far as I understood this question and the first answer here: https://www.reddit.com/r/rust/comments/kbyb6z/what_is_the_purpose_of_reprtransparent/

Same layout does not really suffice, it must be `#[repr(transparent)]` since the layout is with single field structs already the same. But you handled it in the correct way, but if you update some code, maybe you could update the comment. Or maybe I misunderstood again the layout/transparent thingy.